src/AWSCredentials.jl

```diff
@@ -0,0 +1,611 @@
+    if force_refresh || _will_expire(aws_creds)
+        credential_method = aws_creds.renew
+
+        if credential_method !== nothing
+            new_aws_creds = credential_method()
+
+            new_aws_creds === nothing && error("Can't find AWS credentials!")
+            copyto!(aws_creds, new_aws_creds)
+
+            # Ensure credential_method is not overwritten by the new credentials
+            aws_creds.renew = credential_method
+        end
+    end
+
+    return aws_creds
+end
+
+
+function _will_expire(aws_creds::AWSCredentials)
+    return now(UTC) >= aws_creds.expiry - Minute(5)
+end
+
+
+"""
+    _ec2_metadata(metadata_endpoint::String) -> Union{String, Nothing}
+
+Retrieve the EC2 meta data from the local AWS endpoint. Return the EC2 metadata request
+body, or `nothing` if not running on an EC2 instance.
+
+# Arguments
+- `metadata_endpoint::String`: AWS internal meta data endpoint to hit
+
+# Throws
+- `StatusError`: If the response status is >= 300
+- `ParsingError`: Invalid HTTP request target
+"""
+function _ec2_metadata(metadata_endpoint::String)
+    try
+        request = @mock http_get(
+            "http://169.254.169.254/latest/meta-data/$metadata_endpoint"
+        )
+
+        return String(request.body)
+    catch e
+        e isa IOError || rethrow(e)
+    end
+
+    return nothing
+end
+
+
+"""
+    ec2_instance_credentials() -> AWSCredentials
+
+Parse the EC2 metadata to retrieve AWSCredentials.
+"""
+function ec2_instance_credentials()
+    info = _ec2_metadata("iam/info")
+    info = JSON.parse(info)
+
+    name = _ec2_metadata("iam/security-credentials/")
+    creds = _ec2_metadata("iam/security-credentials/$name")
+    new_creds = JSON.parse(creds)
+
+    expiry = DateTime(strip(new_creds["Expiration"], 'Z'))
+
+    return AWSCredentials(
+        new_creds["AccessKeyId"],
+        new_creds["SecretAccessKey"],
+        new_creds["Token"],
+        info["InstanceProfileArn"];
+        expiry=expiry,
+        renew=ec2_instance_credentials
+    )
+end
+
+
+"""
+    ecs_instance_credentials() -> Union{AWSCredential, Nothing}
+
+Retrieve credentials from the local endpoint. Return `nothing` if not running on an ECS
+instance.
+
+More information can be found at:
+https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html
+
+# Returns
+- `AWSCredentials`: AWSCredentials from `ECS` credentials URI, `nothing` if the Env Var is
+    not set (not running on an ECS container instance)
+
+# Throws
+- `StatusError`: If the response status is >= 300
+- `ParsingError`: Invalid HTTP request target
+"""
+function ecs_instance_credentials()
+    if !haskey(ENV, "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
+        return nothing
+    end
+
+    uri = ENV["AWS_CONTAINER_CREDENTIALS_RELATIVE_URI"]
+
+    response = @mock http_get("http://169.254.170.2$uri")
+    new_creds = String(response.body)
+    new_creds = JSON.parse(new_creds)
+
+    expiry = DateTime(strip(new_creds["Expiration"], 'Z'))
```

Comment:
Doesn't really matter:
```suggestion
    expiry = DateTime(rstrip(new_creds["Expiration"], 'Z'))
```